lib/pure/strutils.nim

```diff
@@ -1779,17 +1779,7 @@ func initSkipTable*(a: var SkipTable, sub: string) {.rtl,
     extern: "nsuInitSkipTable".} =
   ## Preprocess table `a` for `sub`.
   let m = len(sub)
-  var i = 0
-  while i <= 0xff-7:
-    a[chr(i + 0)] = m
-    a[chr(i + 1)] = m
-    a[chr(i + 2)] = m
-    a[chr(i + 3)] = m
-    a[chr(i + 4)] = m
-    a[chr(i + 5)] = m
-    a[chr(i + 6)] = m
-    a[chr(i + 7)] = m
-    i += 8
+  algorithm.fill[int](a, m)
```

Comment:
```suggestion
  fill(a, m)
```

is loop unrolling (as before this PR) still happening with fill with optimizations on?
or is previous code a case of premature optimzation?

Reply:
When not using `algorithm.fill`, the generated C code looks like
```c
N_LIB_PRIVATE N_NIMCALL(void, nsuInitSkipTable)(NI* a, NimStringDesc* sub) {
    NI m;
    NI i;
    m = (sub ? sub->Sup.len : 0);
    i = ((NI) 0);
    {
        while (1) {
            if (!(i <= ((NI) 248))) goto LA2;
            a[(((NU8)(((NIM_CHAR) (((NI) ((NI)(i + ((NI) 0)))))))))- 0] = m;
            a[(((NU8)(((NIM_CHAR) (((NI) ((NI)(i + ((NI) 1)))))))))- 0] = m;
            a[(((NU8)(((NIM_CHAR) (((NI) ((NI)(i + ((NI) 2)))))))))- 0] = m;
            a[(((NU8)(((NIM_CHAR) (((NI) ((NI)(i + ((NI) 3)))))))))- 0] = m;
            a[(((NU8)(((NIM_CHAR) (((NI) ((NI)(i + ((NI) 4)))))))))- 0] = m;
            a[(((NU8)(((NIM_CHAR) (((NI) ((NI)(i + ((NI) 5)))))))))- 0] = m;
            a[(((NU8)(((NIM_CHAR) (((NI) ((NI)(i + ((NI) 6)))))))))- 0] = m;
            a[(((NU8)(((NIM_CHAR) (((NI) ((NI)(i + ((NI) 7)))))))))- 0] = m;
            i += ((NI) 8);
        } LA2: ;
    }
    /// rest of the method...
```

When using `algorithm.fill`, the generated C code looks like
```c
N_LIB_PRIVATE N_NIMCALL(void, fill_pureZstrutils_1383)(NI* a, NI aLen_0, NI value) {
    NI xX60gensym163_;
    xX60gensym163_ = ((NI) 0);
    {
        while (1) {
            if (!(xX60gensym163_ <= (aLen_0-1))) goto LA2;
            a[xX60gensym163_] = value;
            xX60gensym163_ += ((NI) 1);
        } LA2: ;
    }
}
N_LIB_PRIVATE N_NIMCALL(void, nsuInitSkipTable)(NI* a, NimStringDesc* sub) {
    NI m;
    m = (sub ? sub->Sup.len : 0);
    fill_pureZstrutils_1383(a, 256, m);
    // rest of the method...
```

So while `algorithm.fill` does not generated unrolled C code, I think a modern C compiler would be able to correctly optimize the code generated by `algorithm.fill`.

These were compiled with `-d:danger --gc:none --checks:off --opt:speed`.

Reply:
that's what i thought too (premature optimization); and in fact I'm now seeing memset can be 30x slower that a loop (when dealing with chars); can share details if needed (refs D20210603T142208)

Reply:
I am certainly curious about the details on that. What does the identifier "D20210603T142208" refer to?

Reply:
=> see https://github.com/nim-lang/Nim/pull/18138#issuecomment-854936534